src/rosdiscover/observer/observer.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+__all__ = ("Observer",)
+import contextlib
+from abc import ABC, abstractmethod
+from typing import Dict, Iterator
+import roswire
+from roswire import AppInstance, ROSVersion
+
+from ..config import Config
+from ..interpreter import NodeContext, SystemSummary
+
+
+class Observer(ABC):
+
+    @classmethod
+    @contextlib.contextmanager
+    def for_container(cls,
+                      container: str,
+                      config: Config,
+                      ) -> Iterator['Observer']:
+        """Constructs and interpreter for a given running container"""
+        rsw = roswire.ROSWire()
+        app = rsw.app(config.image, config.sources)
+        instance = app.attach(container, require_description=True)
+        if app.description.distribution.ros == ROSVersion.ROS1:
+            from .ros1 import ROS1Observer
+            yield ROS1Observer(instance, config)
+        else:
+            from .ros2 import ROS2Observer
+            yield ROS2Observer(instance, config)
+
+    def __init__(self, app: AppInstance, config: Config):
+        self._app_instance = app
+        self._config = config
+        self._nodes: Dict[str, NodeContext] = {}
+
+    def summarise(self):
```

Comment:
Return type annotation

Reply:
I'm not sure that this method belongs to the public API of `Observer`? The only API interaction that I can see is with an `.observe()` method that returns the instantaneous state of the associated system. Everything else is an implementation detail. (That's not to say that this method isn't useful, just that it might make more sense to change it to a protected method.)

Also, fitting with the above comment on statelessness, it might make sense to update this method to accept a `nodes` parameter rather than rely on `self._nodes`.

Reply:
I was thinking that in the future we might want to be able to call this, which is why I made the class stateful.